src/lib_json/json_reader.cpp

```diff
@@ -1417,6 +1417,8 @@ bool OurReader::readString() {
   Char c = 0;
   while (current_ != end_) {
     c = getNextChar();
+    if (c < 0)
```

Comment:
This check isn't make any sense, type `char` will not less than zero.

Reply:
Are you sure? What about `unsigned char`?

Actually, there was an error report without that. And it works fine with it.

Reply:
> This check isn't make any sense, type `char` will not less than zero.


the `Reader::Char` typedef is just char. It is implementation defined whether char is signed or unsigned. I agree it's concerning to see it compared with c<0. On some platforms it's impossible. Others it's fine. So it's not portable code.

Note that char, signed char, and unsigned char are three distinct types in the C++ type system.

(chiming in from myphone)

Reply:
In GBK or GB2312, Chinese characters are composed of two BYTEs, A BYTE greater than 127 followed by any BYTE.

BYTE is unsigned char.
`BYTE` than 127 is a `char` less than 0.

Reply:
> In GBK or GB2312, Chinese characters are composed of two BYTEs, A BYTE greater than 127 followed by any BYTE.
> 
> 
> 
> BYTE is unsigned char.
> 
> `BYTE` than 127 is a `char` less than 0.
> 
> 

Ok there's your problem. We don't decode GBK. That's not a valid JSON string encoding. You will have to convert to UTF-8 before feeding to a JsonCpp Reader.

BYTE may have gone out as an unsigned char, but isn't unsigned char unless the receiver doing the decoding decides to look at it as an unsigned char.

Reply:
Yes, you are very correct! Thank you!